openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,333 @@
+        preference.save()
+
+        if settings.FEATURES.get('SEGMENT_IO_LMS') and settings.SEGMENT_IO_LMS_KEY:
+            _track_update_email_opt_in(user.id, org, optin)
+
+    except IntegrityError as err:
+        log.warn(u"Could not update organization wide preference due to IntegrityError: {}".format(err.message))
+
+
+def _track_update_email_opt_in(user_id, organization, opt_in):
+    """Track an email opt-in preference change.
+
+    Arguments:
+        user_id (str): The ID of the user making the preference change.
+        organization (str): The organization whose emails are being opted into or out of by the user.
+        opt_in (Boolean): Whether the user has chosen to opt-in to emails from the organization.
+
+    Returns:
+        None
+
+    """
+    event_name = 'edx.bi.user.org_email.opted_in' if opt_in else 'edx.bi.user.org_email.opted_out'
+    tracking_context = tracker.get_tracker().resolve_context()
+
+    analytics.track(
+        user_id,
+        event_name,
+        {
+            'category': 'communication',
+            'label': organization
+        },
+        context={
+            'Google Analytics': {
+                'clientId': tracking_context.get('client_id')
+            }
+        }
+    )
+
+
+def _get_user(requesting_user, username=None, allow_staff=False):
+    """
+    Helper method to return the user for a given username.
+    If username is not provided, requesting_user.username is assumed.
+    """
+    if username is None:
+        username = requesting_user.username
+
+    try:
+        existing_user = User.objects.get(username=username)
+    except ObjectDoesNotExist:
+        raise UserNotFound()
+
+    if requesting_user.username != username:
+        if not requesting_user.is_staff or not allow_staff:
+            raise UserNotAuthorized()
+
+    return existing_user
+
+
+def _validate_user_preference(user, preference_key, preference_value):
+    """Helper method to ensure that a user preference is valid.
+
+    Arguments:
+        user (User): The user whose preference is being validated.
+        preference_key (string): The key for the user preference.
+        preference_value (string): The value to be stored. Non-string values will be converted to strings.
+
+    Raises:
+        PreferenceValidationError: the supplied key and/or value for a user preference are invalid.
+    """
+    try:
+        user_preference = UserPreference.objects.get(user=user, key=preference_key)
+    except ObjectDoesNotExist:
+        user_preference = None
+    new_data = {
+        "user": user.id,
+        "key": preference_key,
+        "value": preference_value,
+    }
+    if user_preference:
+        serializer = RawUserPreferenceSerializer(user_preference, data=new_data)
+    else:
+        serializer = RawUserPreferenceSerializer(data=new_data)
+    if not serializer.is_valid():
+        error_message = u"The user preference has the following errors: {errors}".format(errors=serializer.errors)
```

Comment:
serializer.errors is in a pretty ugly format. Did you see what this actually ends up looking like?

It would be good if we are consistent with how validation messages are returned between accounts and preferences. This is what I did:

def _add_serializer_errors(update, serializer, field_errors):
    """
    Helper method that adds any validation errors that are present in the serializer to
    the supplied field_errors dict.
    """
    if not serializer.is_valid():
        errors = serializer.errors
        for key, value in errors.iteritems():
            if isinstance(value, list) and len(value) > 0:
                developer_message = value[0]
            else:
                developer_message = "Invalid value: {field_value}'".format(field_value=update[key])
            field_errors[key] = {
                "developer_message": developer_message,
                "user_message": _("Value '{field_value}' is not valid for field '{field_name}'.".format(
                    field_value=update[key], field_name=key)
                )
            }

    return field_errors

Perhaps this can be generalized just slightly and used for both? We would want to put the method in a common location, not introduce dependency from one api file to the other.